tests/components/zimi/test_config_flow.py

```diff
@@ -0,0 +1,318 @@
+        ControlPointInvalidHostError
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: INPUT_HOST,
+            CONF_PORT: INPUT_PORT,
+        },
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": "invalid_host"}
+
+    # Second attempt succeeds
+    discovery_mock.return_value.validate_connection.side_effect = None
+    discovery_mock.return_value.validate_connection.return_value = (
+        ControlPointDescription(host=INPUT_HOST, port=INPUT_PORT, mac=INPUT_MAC)
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: INPUT_HOST,
+            CONF_PORT: INPUT_PORT,
+        },
+    )
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == f"ZIMI Controller ({INPUT_HOST}:{INPUT_PORT})"
+    assert result["data"] == {
+        "host": INPUT_HOST,
+        "port": INPUT_PORT,
+        "mac": format_mac(INPUT_MAC),
+    }
+
+
+@pytest.mark.parametrize(
+    ("side_effect", "error_expected"),
+    [
+        (
+            ControlPointConnectionRefusedError,
+            {"base": "connection_refused"},
+        ),
+        (
+            ControlPointTimeoutError,
+            {"base": "timeout"},
+        ),
+    ],
+)
+async def test_finish_manual_socket_errors(
+    hass: HomeAssistant,
+    discovery_mock: MagicMock,
+    side_effect: Exception,
+    error_expected: dict,
+) -> None:
+    """Test finish form transitions via socket errors to creation."""
+
+    discovery_mock.discovers.side_effect = ControlPointError("Discovery failed")
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    # First attempt fails with socket errors
+    discovery_mock.return_value.validate_connection.side_effect = side_effect
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: INPUT_HOST,
+            CONF_PORT: INPUT_PORT,
+        },
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == error_expected
+
+    # Second attempt succeeds
+    discovery_mock.return_value.validate_connection.side_effect = None
+    discovery_mock.return_value.validate_connection.return_value = (
+        ControlPointDescription(host=INPUT_HOST, port=INPUT_PORT, mac=INPUT_MAC)
+    )
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_HOST: INPUT_HOST,
+            CONF_PORT: INPUT_PORT,
+        },
+    )
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == f"ZIMI Controller ({INPUT_HOST}:{INPUT_PORT})"
+    assert result["data"] == {
+        "host": INPUT_HOST,
+        "port": INPUT_PORT,
+        "mac": format_mac(INPUT_MAC),
```

Comment:
I'm missing a test of already configured abort reason. Add a MockConfigEntry to hass and then start the flow.